Support graceful agent interruption

pakrym-oai opened the pull request.

pakrym-oai committed:
Support gracefull agent interruption

pakrym-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -1170,19 +1172,6 @@ impl Session {
         self.abort_all_tasks(TurnAbortReason::Interrupted).await;
     }
 
-    fn interrupt_task_sync(&self) {
```
This was only used for abort on drop an AbortOnDropHandle should do the same but easier. @jif-oai please confirm.

pakrym-oai committed:
state/turn.rs, tasks/mod.rs: rename RunningTask::_handle to handle and update references

jif-oai commented on codex-rs/core/src/or_cancel.rs:
```diff
@@ -0,0 +1,31 @@
+use async_trait::async_trait;
```
Can we move this in a small crate? This is cool and we may need it in the futur

jif-oai commented on codex-rs/core/src/error.rs:
```diff
@@ -50,6 +50,9 @@ pub enum SandboxErr {
 
 #[derive(Error, Debug)]
 pub enum CodexErr {
+    #[error("turn aborted")]
```
Can you derive it from the cancel error to avoid all the map_err()?

jif-oai commented on codex-rs/core/src/tasks/mod.rs:
```diff
@@ -143,14 +165,24 @@ impl Session {
+            }
+        }
+
+        task._handle.abort();
```
You should only abort if the task did not closed successfully. Otherwise you might have a race condition

pakrym-oai replied:
Is there a race condition here? I thought abort is idempotent.

pakrym-oai committed:
core: add test-log dependency and improve abort task tests

Add `test-log` as a dev-dependency for logging in tests. Refactor abort task tests for better coverage, multi-threaded execution, and reliable event timing. Enhance `NeverEndingTask` to optionally await cancellation token. Fix task completion event emission to avoid reporting finished after cancellation. Update Cargo.toml and lock files accordingly.

pakrym-oai commented on codex-rs/core/src/tasks/mod.rs:
```diff
@@ -23,6 +29,8 @@ pub(crate) use compact::CompactTask;
 pub(crate) use regular::RegularTask;
 pub(crate) use review::ReviewTask;
 
+const GRACEFULL_INTERRUPTION_TIMEOUT_MS: u64 = 100;
```
We should be using `or_cancel` where it matters and immidiately terminating those network awaits.

pakrym-oai committed:
core: simplify cancellation error handling using `From<CancelErr> for CodexErr`

pakrym-oai committed:
core: fix typo in comment ('repored' -> 'reported')

pakrym-oai committed:
core: move OrCancelExt, CancelErr, and implementation to new codex-async-utils crate

- Add codex-async-utils crate and dependency to workspace and core
- Move or_cancel.rs contents (OrCancelExt trait and CancelErr) from core to codex-async-utils
- Update all uses and imports to reference codex_async_utils::{OrCancelExt, CancelErr}
- Remove or_cancel.rs from core

pakrym-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -2126,6 +2119,7 @@ async fn try_run_turn(
     sub_id: &str,
     prompt: &Prompt,
     task_kind: TaskKind,
+    cancellation_token: CancellationToken,
```
I'll move it into TokenContext in a followup.
